pennylane/compiler/python_compiler/transforms/mbqc/outline_state_evolution.py

```diff
@@ -0,0 +1,465 @@
+                    del qubit_to_reg_idx[op.in_qubit]
+                case quantum.CustomOp():
+                    # To update the qubit_to_reg_idx map for the return type.
+                    for i, qb in enumerate(chain(op.in_qubits, op.in_ctrl_qubits)):
+                        qubit_to_reg_idx[op.results[i]] = qubit_to_reg_idx[qb]
+                        del qubit_to_reg_idx[qb]
+                case quantum.InsertOp():
+                    if op.idx_attr and qubit_to_reg_idx[op.qubit] is not op.idx_attr:
+                        raise ValueError("op.qubit should be op.idx_attr.")
+                    del qubit_to_reg_idx[op.qubit]
+                    # update register since it might have changed
+                    op.operands = (current_reg, op.idx, op.qubit)
+                    current_reg = op.out_qreg
+
+                case _ if (
+                    isinstance(
+                        op,
+                        (
+                            quantum.ComputationalBasisOp,
+                            quantum.NamedObsOp,
+                            quantum.HamiltonianOp,
+                            quantum.TensorOp,
+                        ),
+                    )
+                    and not terminal_boundary_op
+                ):
+                    insert_ops = set()
+
+                    # Insert all qubits recorded in the qubit_to_reg_idx dict before the
+                    # pre-assumed terminal operations.
+                    rewriter.insertion_point = InsertPoint.before(op)
+                    for qb, idx in qubit_to_reg_idx.items():
+                        insert_op = quantum.InsertOp(current_reg, idx, qb)
+                        rewriter.insert(insert_op)
+                        insert_ops.add(insert_op)
+                        terminal_boundary_op = insert_op
+                        current_reg = insert_op.out_qreg
+
+                    # Add the `"terminal_boundary"` attribute to the last newly added
+                    # `quantum.insert` operation.
+                    if terminal_boundary_op is None:
+                        raise RuntimeError("A terminal_boundary_op op is not found in the circuit.")
+                    terminal_boundary_op.attributes["terminal_boundary"] = builtin.UnitAttr()
+
+                    # extract ops
+                    rewriter.insertion_point = InsertPoint.before(op)
+                    for qb, idx in list(qubit_to_reg_idx.items()):
+                        extract_op = quantum.ExtractOp(current_reg, idx)
+                        rewriter.insert(extract_op)
+                        qb.replace_by_if(
+                            extract_op.qubit, lambda use: use.operation not in insert_ops
+                        )
+                        # update the qubit_to_reg_idx dict
+                        qubit_to_reg_idx[extract_op.qubit] = idx
+                        del qubit_to_reg_idx[qb]
```

Comment:
I wonder if it would be better to express this large body as a separate function(s)?

Reply:
Done